gguf-py/gguf/metadata.py

```diff
@@ -0,0 +1,408 @@
+
+        with open(model_card_path, "r", encoding="utf-8") as f:
+            return cast("dict[str, Any]", frontmatter.load(f))
+
+    @staticmethod
+    def load_hf_parameters(model_path: Optional[Path] = None) -> dict[str, Any]:
+        if model_path is None or not model_path.exists():
+            return {}
+
+        config_path = model_path / "config.json"
+
+        if not config_path.exists():
+            return {}
+
+        with open(config_path, "r", encoding="utf-8") as f:
+            return json.load(f)
+
+    @staticmethod
+    def id_to_title(string):
+        # Convert capitalization into title form unless acronym or version number
+        return ' '.join([w.title() if w.islower() and not re.match(r'^(v\d+(?:\.\d+)*|\d.*)$', w) else w for w in string.strip().replace('-', ' ').split()])
+
+    @staticmethod
+    def get_model_id_components(model_id: Optional[str] = None) -> tuple[str | None, str | None, str | None, str | None, str | None, str | None]:
+        # Huggingface often store model id as '<org>/<model name>'
+        # so let's parse it and apply some heuristics if possible for model name components
+
+        if model_id is None:
+            # model ID missing
+            return None, None, None, None, None, None
+
+        if ' ' in model_id:
+            # model ID is actually a normal human sentence
+            # which means its most likely a normal model name only
+            # not part of the hugging face naming standard, but whatever
+            return model_id, None, None, None, None, None
+
+        if '/' in model_id:
+            # model ID (huggingface style)
+            org_component, model_full_name_component = model_id.split('/', 1)
+        else:
+            # model ID but missing org components
+            org_component, model_full_name_component = None, model_id
+
+        # Check if we erroneously matched against './' or '../' etc...
+        if org_component is not None and org_component[0] == '.':
+            org_component = None
+
+        # Regular expression to extract model name components
+        # Heuristic to match against cases such as 'Mixtral-8x7B-Instruct-v0.1' or 'Codestral-22B-v0.1'
+        regex_match = re.compile(r'^(?P<basename>[A-Za-z0-9\s]*(?:(?:-(?:(?:[A-Za-z\s][A-Za-z0-9\s]*)|(?:[0-9\s]*)))*))'
+                                 r'(?:-(?P<size_label>(?:\d+x)?\d+[A-Za-z]+)(?:-(?P<finetune>[A-Za-z0-9\s-]+))?)?'
+                                 r'(?:-(?P<version>v\d+(?:\.\d+)*))?$').match(model_full_name_component)
+
+        if not regex_match:
+            return model_full_name_component, org_component, None, None, None, None
+
+        components = regex_match.groupdict()
+        basename = components.get("basename")
+        finetune = components.get("finetune")
+        version = components.get("version")
+        size_label = components.get("size_label")
+
+        return model_full_name_component, org_component, basename, finetune, version, size_label
```

Comment:
This doesn't always work:

```python3
>>> gguf.Metadata.get_model_id_components("Qwen2-57B-A14B-Instruct")
('Qwen2-57B-A14B-Instruct', None, 'Qwen2', 'A14B-Instruct', None, '57B')
```

`A14B` should be part of the size label.

---

```python
>>> gguf.Metadata.get_model_id_components("Qwen1.5-MoE-A2.7B-Chat")
('Qwen1.5-MoE-A2.7B-Chat', None, None, None, None, None)
```

Nothing is extracted, but at least nothing is wrongly extracted.

---

```python3
>>> gguf.Metadata.get_model_id_components("microsoft/Phi-3-mini-4k-instruct")
('Phi-3-mini-4k-instruct', 'microsoft', 'Phi-3-mini', 'instruct', None, '4k')
```

Pretty sure Phi-3-mini has more than 4k parameters